google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/JdbcDriver.java

```diff
@@ -16,12 +16,225 @@
- * class in a separate PR.
+ * JDBC {@link Driver} for Google Cloud Spanner.
+ *
+ * Usage:
+ *
+ * <pre>
+ * <!--SNIPPET {@link JdbcDriver} usage-->
+ * {@code
+ * String url = "jdbc:cloudspanner:/projects/my_project_id/"
+ *            + "instances/my_instance_id/databases/my_database_name?"
+ *            + "credentials=/home/cloudspanner-keys/my-key.json;autocommit=false";
+ * try (Connection connection = DriverManager.getConnection(url)) {
+ *   try(ResultSet rs = connection.createStatement().executeQuery("SELECT SingerId, AlbumId, MarketingBudget FROM Albums")) {
+ *     while(rs.next()) {
+ *       // do something
+ *     }
+ *   }
+ * }
+ * }
+ * <!--SNIPPET {@link JdbcDriver} usage-->
+ * </pre>
+ *
+ * The connection that is returned will implement the interface {@link CloudSpannerJdbcConnection}.
+ * The JDBC connection URL must be specified in the following format:
+ *
+ * <pre>
+ * jdbc:cloudspanner:[//host[:port]]/projects/project-id[/instances/instance-id[/databases/database-name]][\?property-name=property-value[;property-name=property-value]*]?
+ * </pre>
+ *
+ * The property-value strings should be url-encoded.
+ *
+ * The project-id part of the URI may be filled with the placeholder DEFAULT_PROJECT_ID. This
+ * placeholder will be replaced by the default project id of the environment that is requesting a
+ * connection.
+ *
+ * The supported properties are:
+ * <ul>
+ * <li>credentials (String): URL for the credentials file to use for the connection. If you do not
+ * specify any credentials at all, the default credentials of the environment as returned by
+ * {@link GoogleCredentials#getApplicationDefault()} will be used.</li>
+ * <li>autocommit (boolean): Sets the initial autocommit mode for the connection. Default is
+ * true.</li>
+ * <li>readonly (boolean): Sets the initial readonly mode for the connection. Default is false.</li>
+ * <li>retryAbortsInternally (boolean): Sets the initial retryAbortsInternally mode for the
+ * connection. Default is true. @see
+ * {@link com.google.cloud.spanner.jdbc.CloudSpannerJdbcConnection#setRetryAbortsInternally(boolean)}
+ * for more information.</li>
+ * </ul>
  */
-class JdbcDriver {
+public class JdbcDriver implements Driver {
+  private static final String JDBC_API_CLIENT_LIB_TOKEN = "sp-jdbc";
+  static final int MAJOR_VERSION = 1;
+  static final int MINOR_VERSION = 0;
+  private static final String JDBC_URL_FORMAT =
+      "jdbc:" + ConnectionOptions.Builder.SPANNER_URI_FORMAT;
+  private static final Pattern URL_PATTERN = Pattern.compile(JDBC_URL_FORMAT);
+
+  @InternalApi
   public static String getClientLibToken() {
-    return "sp-jdbc";
+    return JDBC_API_CLIENT_LIB_TOKEN;
+  }
+
+  static {
+    try {
+      register();
+    } catch (SQLException e) {
+      java.sql.DriverManager.println("Registering driver failed: " + e.getMessage());
+    }
+  }
+  private static JdbcDriver registeredDriver;
+
+  static void register() throws SQLException {
```

Comment:
Sorry, one last question. How come this is different than the class loading done in JdbcDataSource?

Reply:
The JdbcDataSource is not responsible for registering the JDBC driver. That is the responsibility of the driver itself, and is done automatically when the driver class (JdbcDriver) is loaded. In most newer environments this is done automatically thanks to the service registration in META-INF/services/java.sqlDriver. The JdbcDataSource does a manual load of the driver class to be 100% certain that the driver is loaded and registered, although in most cases it will be redundant.